### DISCUSSION TITLE
Usability enhancement: Returning to query editor after running a query

### DISCUSSION THREAD

**Discussion by ml31415** [tconbeer/harlequin#507] [Q&A, Troubleshooting, Support] [answered]
Right now, the focus jumps from the editor to the result window after running a query. Which is probably a reasonable default for most use cases.

Unfortunately, returning to the query editor after this forceful jump to the result window is a bit annoying. If I didn't overlook something, it currently requires leaving the comfortable base line of your keyboard and hitting F2 or grabbing the mouse. It would be very convenient for any workflow, that requires editing the query several times, to have another ctrl+something keyboard binding for this. Ctrl+k might be a reasonable choice, as ctrl+q and ctrl+e are already taken, and it would be near the ctrl+j, which was likely just used for running the query before.

**Comment by tconbeer** [comment]
Thanks for the feedback. I personally like <kbd>F2</kbd> for this, but I understand you have different preferences.

There aren't many key bindings to spare for the editor, but there are more options from the results viewer. <kbd>ctrl+k</kbd> could work, although that might be confusingly close to <kbd>k</kbd>, which switches to the next results viewer tab (if more than one query executed). We could also do another single key, like <kbd>;</kbd>. Interested to hear more thoughts on this, and/or do some research on what the other SQL IDEs use (e.g., datagrip, dbeaver, etc.). Also probably relates to #422 and #371 

We do need to reserve some bindings for sorting and filtering the Results Viewer, but otherwise the taken ones are:

### Results Viewer Bindings

- ctrl+c: Copy selected data to the clipboard.

#### Switching Tabs

- j: Switch to the previous tab.
- k: Switch to the next tab.

#### Moving the Cursor

- up,down,left,right: Move the cursor one cell.
- tab: Move the cursor to the next cell in the table.
- shift+tab: Move the cursor to the previous cell in the table.
- home, ctrl+up: Move the cursor to the top of the current column.
- end, ctrl+down: Move the cursor to the bottom of the current column.
- ctrl+left: Move the cursor to the start of the current row.
- ctrl+right: Move the cursor to the end of the current row.
- ctrl+home: Move the cursor to the start of the table.
- ctrl+end: Move the cursor to the end of the table.
- PgUp: Move the cursor up one screen.
- PgDn: Move the cursor down one screen.
- shift+[any]: Select range while moving the cursor.

**Reply by tconbeer** [reply]
In Datagrip, `Esc` from any "window" focuses the active editor again.
I can't tell if DBeaver has something similar.

**Comment by ml31415** [comment]
Big thanks for having a look into this!

It might also be an option, to make ctrl+j not a global "Run query" hotkey, but only run query on ctrl+j, when the editor is focused. And use the same ctrl+j to jump back to the editor when the results are focused. This would save wasting another hotkey.

For the F-keys: Since most keyboards started to use them as media keys as well, and it requires the FN-modifier pressed to reach them, it got even more unpleasant to use them. I'd suggest to reserve them mainly for rarely used actions, opening help etc.

**Reply by tconbeer** [reply, marked as answer]
This is now configurable, by the way. Use `harlequin --keys` to generate a new keymap that sets the `Results Viewer: Focus Query Editor` action.

**Comment by mahiki** [comment]
I usually only look at the results, and dont need to take action on those results, and so hitting a key to get back to the editor is friction. I understand others prefer this default, it seems like window switching should be a configuration.

But on that topic, there doesn't seem to be an application level configuration file? The .harlequin.toml sets up connection info only AFAICT.

**Reply by tconbeer** [reply]
I'm open to this being configurable via the TOML file if you want to open a PR for that.

**Reply by ml31415** [reply]
I wonder, for how many people it is the default, to perfectly get a query straight and then instantly want to browse. At least for me it's quite often tinkering on the SQL several times and only then get into the data, if the result set is narrowed down good enough.
